src/openms/include/OpenMS/KERNEL/RangeManager.h

```diff
@@ -28,181 +28,574 @@
+    {
+      return max_;
+    }
+    ///@}
+
+    /// extend the range such that it includes the given @p value
+    void extendIntensity(const double value)
+    {
+      extend(value);
+    }
+  };
+
+  OPENMS_DLLAPI std::ostream& operator<<(std::ostream& out, const RangeIntensity& range);
+
+  struct OPENMS_DLLAPI RangeMobility : public RangeBase {
+
+      const static MSDim DIM = MSDim::IM;
+
+      RangeMobility() = default;
+      RangeMobility(const double min, const double max) :
+          RangeBase(min, max)
       {
-        return;
       }
 
-      PositionType min = pos_range_.minPosition();
-      PositionType max = pos_range_.maxPosition();
+      /** @name Accessors for min and max
+      
+          We use accessors, to keep range consistent (i.e. ensure that min <= max)
+      */
+      ///@{
+
+      /// sets the minimum (and the maximum, if uninitialized)
+      void setMinMobility(const double min)
+      {
+        setMin(min);
+      }
 
-      double it_min = int_range_.minPosition()[0];
-      double it_max = int_range_.maxPosition()[0];
+      /// sets the maximum (and the minimum, if uninitialized)
+      void setMaxMobility(const double max)
+      {
+        setMax(max);
+      }
 
-      for (PeakIteratorType it = begin; it != end; ++it)
+      /// only useful if isEmpty() returns false
+      double getMinMobility() const
       {
-        //update position
-        for (UInt i = 0; i < D; ++i)
+        return min_;
+      }
+
+      /// only useful if isEmpty() returns false
+      double getMaxMobility() const
+      {
+        return max_;
+      }
+      ///@}
+
+      /// extend the range such that it includes the given @p value
+      void extendMobility(const double value)
+      {
+        extend(value);
+      }
+    };
+
+  OPENMS_DLLAPI std::ostream& operator<<(std::ostream& out, const RangeMobility& range);
+  
+  /// Enum listing state of dimensions (RangeBases)
+  enum class HasRangeType
+  {
+    ALL, ///< all dimensions are filled
+    SOME,///< some dimensions are empty, some are filled
+    NONE ///< all dimensions are empty (=cleared)
+  };
+
+  /**
+    @brief Handles the management of a multidimensional range, e.g. RangeMZ and RangeIntensity for spectra.
+
+    Instanciate it with the dimensions which are supported/required, e.g.
+    <tt>RangeManager<RangeRT, RangeMZ> range_spec</tt> for a spectrum and use the stronly typed features, such as
+    range_spec.getMaxRT()/setMaxRT(500.0) or range_spec.extend(RangeMZ{100, 1500});
+
+    Use RangeManagerContainer as a base class for all peak and feature containers like MSSpectrum, MSExperiment and FeatureMap.
+
+  */
+  template<typename... RangeBases>
+  class RangeManager : public RangeBases...
+  {
+  public:
+    // rule of 0 -- no need for a virtual d'tor or anything fancy
+    // ...
+
+    bool operator==(const RangeManager& rhs) const
+    {
+      bool equal = true;
+      for_each_base([&](auto* base) {
```

Comment:
can you add some comment here. for the for each base and the decay_t / decltype as we never used that before in the code base

Reply:
sure, can do